lib/mongo/server_selector/selectable.rb

```diff
@@ -212,9 +220,48 @@ def match_tag_sets(candidates)
         matches || []
       end
 
-      def validate_tag_sets!(tag_sets)
-        if !tag_sets.all? { |set| set.empty? } && !tags_allowed?
-          raise Error::InvalidServerPreference.new(name)
+      def filter_stale_servers(candidates, primary = nil)
+        return candidates unless @max_staleness
+        max_staleness_ms = @max_staleness * 1000
+
+        if primary
+          candidates.select do |server|
+            validate_max_staleness_support!(server)
+            staleness = (server.last_scan - server.last_write_date) -
+                        (primary.last_scan - primary.last_write_date)  +
+                        (server.heartbeat_frequency * 1000)
+            staleness <= max_staleness_ms
+          end
+        else
+          max_write_date = candidates.collect(&:last_write_date).max
+          candidates.select do |server|
+            validate_max_staleness_support!(server)
+            staleness = max_write_date - server.last_write_date + (server.heartbeat_frequency * 1000)
+            staleness <= max_staleness_ms
+          end
+        end
+      end
+
+      def validate!
+        if !@tag_sets.all? { |set| set.empty? } && !tags_allowed?
+          raise Error::InvalidServerPreference.new(Error::InvalidServerPreference::NO_TAG_SUPPORT)
+        elsif @max_staleness && !max_staleness_allowed?
+          raise Error::InvalidServerPreference.new(Error::InvalidServerPreference::NO_MAX_STALENESS_SUPPORT)
+        end
+      end
+
+      def validate_max_staleness_support!(server)
+        return unless @max_staleness
+        if @max_staleness && !server.features.collation_enabled?
```

Comment:
Rename "collation_enabled" to something more generic, or make it a separate property? I understand why this works but it looks surprising.